### DISCUSSION TITLE
url manipulation in RodauthApp > configure > prefix

### DISCUSSION THREAD

**Discussion by whysthatso** [janko/rodauth-rails#193] [Q&A] [answered]
I'm trying to dynamically change the prefix based on the Accept-Language header, so currently i have 
```
  configure do
    prefix { "/#{ RodauthApp.best_locale_from_request! || I18n.locale}" }
  end
```
but i would need something like this:
```
  configure do
    prefix { "/#{  parse( env.fetch("HTTP_ACCEPT_LANGUAGE") )  || I18n.locale }" }
  end
```
I think i can use the AcceptLanguage gem to parse the header, but i don't have access to env here. it's accessible in the `route` block but not in the `configure` block, which is something i do not fully comprehend. is there anyway to access `env` here or how would one do that?

**Comment by janko** [comment, marked as answer]
The `configure` block is executed immediately, but the individual auth method blocks are executed only when calling Rodauth, which happens during the request. The Rodauth object is initialized with the Roda instance, so you have access to the current request. You can access the env via the `request` object, i.e. `request.env`.

**Reply by whysthatso** [reply]
thank you for the answer, but this abstraction is slightly above my paygrade. could you help me understand in what context the request object exists?

i've been trying to dump request.env like so: `Rails.logger.debug request.env` inside `class RodauthApp < Rodauth::Rails::App`, inside the `configure` block for the prefix, to no avail. Only in the `route` block do i get an output, either with `env` or `request.env`

I understand your first sentence, but i'm failing at: "The Rodauth object is initialized with the Roda instance"

**Reply by janko** [reply]
Inside the `prefix` block you're in the context of the Rodauth object. So, there instead of `env` you need to call `request.env`. Does that work? If not, in what way does it fail?

Inside the `route` block you're in the context of a Roda instance. Anything available though Roda is available in the Rodauth instance. I just wanted to explain why you have access to the request, and the connection between Roda and Rodauth instances.

**Reply by whysthatso** [reply]
now it's clear. and my locale setting based on headers is also working.

it seems i have however misunderstood my own problem which is initially:
redirect / to  a locale path i.e. / => /en

i am still confused as to where to do this. all suggestions i found point to redirects inside the application controller  and the rails routes. i guess that's fair, as i do auth on two distinct paths for my two rodauth configs :main (/client/) and :admin  (/admin/)

so as i understand it, although the request for / first goes through rodauth, it is basically ignored, handed to the rails router, and if i want to then do a redirect to i.e. /en/, i have to do that in rails, as rodauth ever only is going to trigger on the configured subpaths.

okay, i think that is it. sorry for spelling this out so verbosely, but i think it might help anybody else on my level that comes across this.
